### ISSUE TITLE
Comment on a specific line of a diff

### ISSUE THREAD

**Issue by kumar303** [mozilla/addons-code-manager#989]
User story:

> As a code reviewer, I want to be able to enter custom comments for specific code lines in the file and diff viewer.

When viewing a diff, the reviewer can enter a comment a specific line. This will be similar to adding a comment to a file (https://github.com/mozilla/addons-code-manager/issues/112).

The suggested UI flow per [meeting notes](https://docs.google.com/document/d/1y4crrHxhGejb3eZ6HpjtBdDqONJakfwtfjaOqwq-qZo/edit) is roughly this:
- click on a line of a diff to begin entering a comment
- when presented with a list of canned responses, click *other*
- begin entering text
- press save

It's possible that it will be easy to hook into `DiffView`. I originally thought it would be tricky and filed https://github.com/otakustay/react-diff-view/issues/59 which led to https://github.com/otakustay/react-diff-view/#customize-gutter . That approach may not be necessary anymore, though.

**Comment by mirefly** [comment]
I guess it should be allowed to comment on deleted line, but it looks that [comment API](https://addons-server.readthedocs.io/en/latest/topics/api/reviewers.html#drafting-comments) does not support comments on deleted part. Do we need API changes before implementing comments in diff mode?

**Comment by bobsilverberg** [comment]
> I guess it should be allowed to comment on deleted line, but it looks that [comment API](https://addons-server.readthedocs.io/en/latest/topics/api/reviewers.html#drafting-comments) does not support comments on deleted part. Do we need API changes before implementing comments in diff mode?

I have come across the same concern. It looks like the API only supports comments on the "new" file. I have pinged @EnTeQuAk on IRC to discuss this.

**Comment by kumar303** [comment]
We can get this feature started by doing what the API supports: only commenting on newer lines.

**Comment by kumar303** [comment]
I bumped the milestone since we're not pushing this week.

**Comment by ioanarusiczki** [comment]
@kumar303 I verified on dev and perhaps there's more to investigate but so far these are my testing results:

- for these compare files:
https://code.addons-dev.allizom.org/en-US/compare/495710/versions/1541609...1688333/?path=manifest.json
https://code.addons-dev.allizom.org/en-US/compare/495710/versions/1541610...1688369/?path=manifest.json
the old line has the "+" button available and it shows the comment field twice, after saving the comment I see it displayed twice . I think it is explained [above](https://github.com/mozilla/addons-code-manager/issues/989#issuecomment-555331672) the reason why this won't work yet.

![the old file](https://user-images.githubusercontent.com/33448286/69879553-913a8e80-127c-11ea-9c66-3fe1c5344c34.gif)


- when I keep the new version but I change the old one for a comment made on the file I can still see it. This doesn't happen when I don't keep the new version.
I don't think this works as expected. I should not see comments made for other versions.

![the same comments](https://user-images.githubusercontent.com/33448286/69879737-12922100-127d-11ea-905a-610cee5c76ed.gif)

- for this file https://code.addons-dev.allizom.org/en-US/compare/390151/versions/1688191...1688302/?path=manifest.json the old line does not have a "+" comment button.

**Comment by bobsilverberg** [comment]
@ioanarusiczki Thanks for the comments. I will investigate these. Follow-up issues may be required.

**Comment by bobsilverberg** [comment]
The first issue raised by @ioanarusiczki above, relates to the fact that we are showing a line in the diff for changes marked with types of `insert-eofnl` and `delete-eofnl`. The code introduced to allow commenting on lines of a diff only excludes changes with a type of `delete`. I could expand that list to make sure we do not allow commenting on `insert-eofnl` and `delete-eofnl`, but I actually think that we probably shouldn't be displaying those lines in the diff. If you look at https://code.addons-dev.allizom.org/en-US/compare/495710/versions/1541609...1688333/?path=manifest.json, for example, you can see that line `80` appears twice, once as a delete (with the contents of `}`) and once as the `insert-eofnl` change, and we also see an extra line at the bottom of the diff for the `delete-eofnl`.

@kumar303 @EnTeQuAk What do you think?

**Comment by bobsilverberg** [comment]
The second issue raised by @ioanarusiczki is:

>
> * when I keep the new version but I change the old one for a comment made on the file I can still see it. This doesn't happen when I don't keep the new version.
> I don't think this works as expected. I should not see comments made for other versions.
> 

This is a result of the fact that comments are stored for a particular version, regardless of whether they relate to a diff or not. We have a separate discussion about this at https://github.com/mozilla/addons-server/issues/12875. 

Because, for now at least, comments are not created in the context of a diff (i.e., a comparison between two versions), but rather just in the context of the newer version, that means that one will see comments added to version A any time one views a comparison between version A and any other version. This is probably something we need to address, but I don't think there's anything we can do about this until we figure out what we want to do about comments on diffs in general.

@ioanarusiczki Could you please file a separate issue for this one, so we can track it separately?

**Comment by kumar303** [comment]
- re: `insert-eofnl` and `delete-eofnl`, let's make sure we have a separate issue to track this. Maybe we just need to disallow comments on `delete-eofnl` similar to how we do it for `delete` changes? Or maybe disallow commenting on both? At the very least, we need to update [ChangeType](https://github.com/mozilla/addons-code-manager/blob/2e1316294c5a855e7e8fdbe4f783504f3f58b84f/src/%40types/react-diff-view/index.d.ts#L4).
* re: displaying head version comments when changing the base version in a comparison: this is by design. If you are *comparing* a head version to a base version, you should still see comments for the head version.

**Comment by bobsilverberg** [comment]
> * re: `insert-eofnl` and `delete-eofnl`, let's make sure we have a separate issue to track this. Maybe we just need to disallow comments on `delete-eofnl` similar to how we do it for `delete` changes? Or maybe disallow commenting on both? At the very least, we need to update [ChangeType](https://github.com/mozilla/addons-code-manager/blob/2e1316294c5a855e7e8fdbe4f783504f3f58b84f/src/%40types/react-diff-view/index.d.ts#L4).
> 

Thanks @kumar303. I opened https://github.com/mozilla/addons-code-manager/issues/1291.
 
> 
> * re: displaying head version comments when changing the base version in a comparison: this is by design. If you are _comparing_ a head version to a base version, you should still see comments for the head version.

My concern about this is that we will lose context. A comment added to line 1 when comparing version 2 to version 3 might not make sense when comparing version 1 to version 3.

**Comment by kumar303** [comment]
> My concern about this is that we will lose context. A comment added to line 1 when comparing version 2 to version 3 might not make sense when comparing version 1 to version 3.

Why not? If you are comparing version 1 to version 3 your intent is still to comprehend the impact of version 3. 

In any case, this is something we can get feedback from reviewers on once the feature lands. They will be able to use it and tell us what makes sense for their work flows.

**Comment by ioanarusiczki** [comment]
@bobsilverberg @kumar303 

Based on https://github.com/mozilla/addons-server/issues/12875#issuecomment-555120040 and also https://github.com/mozilla/addons-code-manager/issues/989#issuecomment-560448350

This is how it works so far for lines that have not been modified from a version to another:

- On line 1 for `manifest.json` file I made a comment on all the 3 versions of an add-on
https://code.addons-dev.allizom.org/en-US/browse/618459/versions/1689844/
So the Browse pages will show now for each version 3 different comments.
For 5.1.1 is "test", for 5.1.2 is "test 5.1.2" and for "5.1.3" is "test 5.1.3"

The Compare View 5.1.2...5.1.3 will show the "test5.1.3" comment only.
The Compare View 5.1.1...5.12 will show the "test 5.1.2" comment only.

I think so far this is as expected.

- I add a comment on line 2 for `manifest.json` file in compare view 5.1.1...5.1.3
It is "Test 5.1.1. versus 5.1.3"

If I change the base version to 5.1.2 the comment will still be available because the head version was kept 5.1.3
The Browse page for 5.1.1 won't show any comment on line 2.
The Browse page for 5.1.3 will show the new comment made ""Test 5.1.1. versus 5.1.3".

Shouldn't this comment also be available for 5.1.1?

- testing the scenarios above for modified lines won't be any different. It's the same behavior.

**Comment by bobsilverberg** [comment]
> @bobsilverberg @kumar303
> 
> Based on [mozilla/addons-server#12875 (comment)](https://github.com/mozilla/addons-server/issues/12875#issuecomment-555120040) and also [#989 (comment)](https://github.com/mozilla/addons-code-manager/issues/989#issuecomment-560448350)
> 
> This is how it works so far for lines that have not been modified from a version to another:
> 
> * On line 1 for `manifest.json` file I made a comment on all the 3 versions of an add-on
> [code.addons-dev.allizom.org/en-US/browse/618459/versions/1689844](https://code.addons-dev.allizom.org/en-US/browse/618459/versions/1689844/)
> So the Browse pages will show now for each version 3 different comments.
> For 5.1.1 is "test", for 5.1.2 is "test 5.1.2" and for "5.1.3" is "test 5.1.3"
> 
> 
> The Compare View 5.1.2...5.1.3 will show the "test5.1.3" comment only.
> The Compare View 5.1.1...5.12 will show the "test 5.1.2" comment only.
> 
> I think so far this is as expected.

Yes, that all sounds as expected.

> 
> * I add a comment on line 2 for `manifest.json` file in compare view 5.1.1...5.1.3
>       It is "Test 5.1.1. versus 5.1.3"
> 
> 
> If I change the base version to 5.1.2 the comment will still be available because the head version was kept 5.1.3
> The Browse page for 5.1.1 won't show any comment on line 2.
> The Browse page for 5.1.3 will show the new comment made ""Test 5.1.1. versus 5.1.3".
> 
> Shouldn't this comment also be available for 5.1.1?

No, when you add a comment via compare for 5.1.1...5.1.3, the comment only gets attached to 5.1.3, so it will only be available in Browse mode for 5.1.3.

Does that make sense?

**Comment by ioanarusiczki** [comment]
@bobsilverberg A bit confused. 
You asked me to file a followup here https://github.com/mozilla/addons-code-manager/issues/989#issuecomment-560445218 


> This is a result of the fact that comments are stored for a particular version, regardless of whether they relate to a diff or not. We have a separate discussion about this at mozilla/addons-server#12875.

> Because, for now at least, comments are not created in the context of a diff (i.e., a comparison between two versions), but rather just in the context of the newer version, that means that one will see comments added to version A any time one views a comparison between version A and any other version. This is probably something we need to address, but I don't think there's anything we can do about this until we figure out what we want to do about comments on diffs in general.

is that still available after the discussions ?

**Comment by ioanarusiczki** [comment]
As discussed on IRC no other followups are to be filed yet. 
Verified on dev - with FF70(Win10).
